ChattoAdditions/Source/Chat Items/CompoundMessage/Views/CompoundBubbleLayout.swift

```diff
@@ -0,0 +1,105 @@
+//
+// The MIT License (MIT)
+//
+// Copyright (c) 2015-present Badoo Trading Limited.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+public struct CompoundBubbleLayout {
+    public let size: CGSize
+    public let subviewsFrames: [CGRect]
+    public let safeAreaInsets: UIEdgeInsets
+}
+
+public struct CompoundBubbleLayoutProvider {
+
+    public struct Configuration: Hashable {
+
+        fileprivate let layoutProviders: [MessageManualLayoutProviderProtocol]
+        fileprivate let tailWidth: CGFloat
+        fileprivate let isIncoming: Bool
+
+        public init(layoutProviders: [MessageManualLayoutProviderProtocol],
+                    tailWidth: CGFloat,
+                    isIncoming: Bool) {
+            self.layoutProviders = layoutProviders
+            self.tailWidth = tailWidth
+            self.isIncoming = isIncoming
+        }
+
+        public func hash(into hasher: inout Hasher) {
+            hasher.combine(self.layoutProviders.map { $0.asHashable })
+            hasher.combine(self.tailWidth)
+            hasher.combine(self.isIncoming)
+        }
+
+        public static func == (lhs: CompoundBubbleLayoutProvider.Configuration,
+                               rhs: CompoundBubbleLayoutProvider.Configuration) -> Bool {
+            return lhs.layoutProviders.map { $0.asHashable } == rhs.layoutProviders.map { $0.asHashable }
+                && lhs.tailWidth == rhs.tailWidth
+                && lhs.isIncoming == rhs.isIncoming
+        }
+    }
+
+    private let configuration: Configuration
+    private let cache = Cache<CGFloat, CompoundBubbleLayout>()
+
+    public init(configuration: Configuration) {
+        self.configuration = configuration
+    }
+
+    public func makeLayout(forMaxWidth width: CGFloat) -> CompoundBubbleLayout {
```

Comment:
Could it be better to split this function into a factory function and accessor function? `layout(forMaxWidth...)` and `makeLayout(forMaxWidth...)`. The factory function could be called when cache is missing a layout for a given width.